libraries/chain/block_log.cpp

```diff
@@ -31,1490 +21,1610 @@ namespace eosio { namespace chain {
-               if( block_file.is_open() && index_file.is_open() && prune_config && prune_config->vacuum_on_close) {
-                  if(!head) {
-                     //disregard vacuum_on_close size if there isn't even a block and just do it silently anyways
-                     vacuum();
-                  }
-                  else {
-                     const size_t first_data_pos = get_block_pos(first_block_num);
-                     block_file.seek_end(-sizeof(uint32_t));
-                     const size_t last_data_pos = block_file.tellp();
-                     if(last_data_pos - first_data_pos < prune_config->vacuum_on_close) {
-                        ilog("Vacuuming pruned block log");
-                        vacuum();
-                     }
-                  }
-               }
-            }
+      uint32_t                                   ver             = 0;
+      uint32_t                                   first_block_num = 0;
+      std::variant<genesis_state, chain_id_type> chain_context;
 
-            void close() {
-               if( block_file.is_open() )
-                  block_file.close();
-               if( index_file.is_open() )
-                  index_file.close();
-               open_files = false;
-            }
+      uint32_t version() const { return ver & ~detail::pruned_version_flag; }
+      bool     is_currently_pruned() const { return ver & detail::pruned_version_flag; }
 
-            template<typename T>
-            void reset( const T& t, const signed_block_ptr& genesis_block, uint32_t first_block_num );
+      chain_id_type chain_id() const {
+         return std::visit(overloaded{ [](const chain_id_type& id) { return id; },
+                                       [](const genesis_state& state) { return state.compute_chain_id(); } },
+                           chain_context);
+      }
 
-            void remove();
+      constexpr static int nbytes_with_chain_id = // the bytes count when the preamble contains chain_id
+            sizeof(uint32_t) + sizeof(first_block_num) + sizeof(chain_id_type) + sizeof(block_log::npos);
+
+      template <typename Stream>
+      void read_from(Stream& ds, const fc::path& log_path) {
+         ds.read((char*)&ver, sizeof(ver));
+         EOS_ASSERT(version() > 0, block_log_exception, "Block log was not setup properly");
+         EOS_ASSERT(block_log::is_supported_version(version()), block_log_unsupported_version,
+                    "Unsupported version of block log. Block log version is ${version} while code supports version(s) "
+                    "[${min},${max}], log file: ${log}",
+                    ("version", version())("min", block_log::min_supported_version)(
+                          "max", block_log::max_supported_version)("log", log_path.generic_string()));
+
+         first_block_num = 1;
+         if (version() != initial_version) {
+            ds.read((char*)&first_block_num, sizeof(first_block_num));
```

Comment:
Is C++ casting preferable to C?
`ds.read( reinterpret_cast<char*>(&first_block_num), sizeof(first_block_num));`

Reply:
we have byte order checking in line 11